src/Execution/Arguments/UpsertModel.php

```diff
@@ -27,16 +27,24 @@ public function __construct($previous)
      */
     public function __invoke($model, $args)
     {
-        /** @var \Nuwave\Lighthouse\Execution\Arguments\Argument|null $id */
-        $id = $args->arguments['id']
-            ?? $args->arguments[$model->getKeyName()]
-            ?? null;
-
-        if ($id) {
-            $model = $model->newQuery()->find($id->value)
-                ?? $model->newInstance();
-        } else {
-            $model = $model->newInstance();
+        if (
+            $id = $args->arguments['id']
+                ?? $args->arguments[$model->getKeyName()]
+                ?? null
+        ) {
+            if (
+                $existingModel = $model
+                    ->newQuery()
+                    ->find($id->value)
+            ) {
+                // TODO there is a slight lingering bug here. In case the $model already
```

Comment:
If I understand you correctly, I think this is issue is not present, but even before this change, considering the previous code:
```
$id = $args->arguments['id']
    ?? $args->arguments[$model->getKeyName()]
    ?? null;

if ($id) {
    $model = $model->newQuery()->find($id->value)
        ?? $model->newInstance();
} else {
    $model = $model->newInstance();
}
```
When there was an `id` input, and the model exists, `$model = $model->newQuery()->find($id->value)`, just like now.

That said, I don't really know how can we deal with it, since I think it needs a design decision. I even think that this could lead us to undesired effects, like allowing to the client to change the relationships of some models, jus think that an `User` could create a `Comment`, then it can update it and change the `user` relationship to another `User`, passing all responsabilities to that User.

I have added a test hoping to explain it a bit more.

Reply:
Well done, yeah this works fine.